### PR TITLE
Implement NodeAddressing on top of Table[NodeAddress]

### PR DESCRIPTION
**Pull request by joamaki**
This works towards making the cilium-agent responsive to node address changes by moving from global variables and direct use of `netlink.AddrList` to using statedb tables that can be queried and watched for changes. 

Reimplement `NodeAddressing` to query against `Table[NodeAddress]` (derived from `Table[*Device]` which is populated from netlink subscriptions to links, addresses and routes) and `LocalNodeStore` instead of using the pkg/node globals or doing a `netlink.AddrList`. Eventual goal is to remove `NodeAddressing` and instead have the modules query `Table[NodeAddress]` directly and reconcile if it changes.

Remove the `ipv{4,6}NodePortAddrs` globals and the related getters and init, replacing uses of them with either `NodeAddressing` or queries to `Table[NodeAddress]`.

See commit messages for more details.

### CODE REVIEW

File: pkg/datapath/tables/node_addressing.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		return n.IPv4AllocCIDR
+	}
+	return nil
+}
+
+func (a addressFamily) LocalAddresses() (addrs []net.IP, err error) {
+	return a.getNodeAddresses(a.db.ReadTxn(), a.flags), nil
+}
+
+func (a addressFamily) DirectRouting() (int, net.IP, bool) {
```
Comment (joamaki) [inline pkg/datapath/tables/node_addressing.go:81]: This is likely half-implemented with not all uses of `option.Config.DirectRoutingDevice` converted.

**Comment by joamaki** [comment]
Assigning to @bimmlerd to push this forward while I'm on vacation.

File: pkg/datapath/tables/node_addressing.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	ipv4     getFlags = 1 << 0
+	ipv6     getFlags = 1 << 1
+	nodePort getFlags = 1 << 2
+)
+
+func (a addressFamily) getDirectRouting(flags getFlags) (int, net.IP, bool) {
+	deviceName := option.Config.DirectRoutingDevice
+	if deviceName == "" {
+		return 0, nil, false
+	}
```
Comment (thorn3r) [inline pkg/datapath/tables/node_addressing.go]: nit: unused local
```suggestion
	if option.Config.DirectRoutingDevice == "" {
		return 0, nil, false
	}
```

**Review by thorn3r — COMMENTED** [review]
small nit on a 1st pass, I see we're working out some failures/merge conflicts so I'll take another look later

**Comment by bimmlerd** [comment]
Fixed all the non-flakes, please do take a look.

**Comment by bimmlerd** [comment]
CI triage:
- e2e failures seem to be related to https://github.com/cilium/cilium/issues/28694

File: pkg/datapath/loader/loader.go [hunk 3]
```diff
@@ -325,7 +325,7 @@ func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, o
 	// Replace program on cilium_net.
 	if _, err := netlink.LinkByName(defaults.SecondHostDevice); err != nil {
 		log.WithError(err).WithField("device", defaults.SecondHostDevice).Error("Link does not exist")
-		return err
+		return fmt.Errorf("device '%s' not found: %w", defaults.SecondHostDevice, err)
```
Comment (ti-mo) [inline pkg/datapath/loader/loader.go:328]: Nit: this is really unlikely to happen, as we create this device a little bit before this, but I guess that's where it went wrong while coming up with this patch. :)

File: pkg/datapath/loader/hash_test.go [hunk 4]
```diff
@@ -20,7 +27,21 @@ var (
 // TesthashDatapath is done in this package just for easy access to dummy
 // configuration objects.
 func (s *LoaderTestSuite) TesthashDatapath(c *C) {
-	cfg := &config.HeaderfileWriter{}
+	var cfg datapath.ConfigWriter
+	hv := hive.New(
+		cell.Provide(
+			fake.NewNodeAddressing,
+			func() bandwidth.Manager { return &fake.BandwidthManager{} },
```
Comment (ti-mo) [inline pkg/datapath/loader/hash_test.go:34]: Why is the bandwidth manager a requirement here?

**Review by ti-mo — APPROVED** [review]

File: pkg/datapath/tables/node_address.go [hunk 5]
```diff
@@ -291,14 +302,14 @@ func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.
 		if n.AddressScopeMax < unix.RT_SCOPE_LINK {
 			for _, addr := range sortedAddresses(dev.Addrs) {
 				if addr.Scope == unix.RT_SCOPE_LINK {
+					n.Log.Infof("HII2 - adding addr: %v prim: %v np: %v dev: %v", addr.Addr, !addr.Secondary, false, dev.Name)
```
Comment (thorn3r) [inline pkg/datapath/tables/node_address.go]: is this leftover from debugging? If not, maybe something more descriptive (HII2 doesn't seem very descriptive without looking at the code)

File: pkg/datapath/tables/node_address.go [hunk 6]
```diff
@@ -302,7 +302,6 @@ func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.
 		if n.AddressScopeMax < unix.RT_SCOPE_LINK {
 			for _, addr := range sortedAddresses(dev.Addrs) {
 				if addr.Scope == unix.RT_SCOPE_LINK {
-					n.Log.Infof("HII2 - adding addr: %v prim: %v np: %v dev: %v", addr.Addr, !addr.Secondary, false, dev.Name)
 					addrs.Insert(NodeAddress{
```
Comment (thorn3r) [inline pkg/datapath/tables/node_address.go:305]: ah, I see they are removed here in a later commit

**Review by thorn3r — APPROVED** [review]
Nice work :+1:

**Review by aditighag — APPROVED** [review]
Approving for pkg/endpoint changes.

File: daemon/cmd/daemon.go [hunk 7]
```diff
@@ -1129,13 +1135,9 @@ func (d *Daemon) ReloadOnDeviceChange(devices []string) {
 	}
 
 	if option.Config.EnableNodePort {
-		if err := node.InitNodePortAddrs(devices, option.Config.LBDevInheritIPAddr); err != nil {
```
Comment (ysksuzuki) [inline daemon/cmd/daemon.go:1132]: `LBDevInheritIPAddr` is no longer referenced. Does this change ignore  `bpf-lb-dev-ip-addr-inherit` flag users specified?

Comment (joamaki) [reply daemon/cmd/daemon.go:1132]: Will fix this in follow-up PR once we've figured out the right way to do it.

**Pull request merged**